File: Python/solutions/18-greatest-common-factor.py
# Write a method that takes in two numbers. Return the greatest
# integer that evenly divides both numbers. You may wish to use the
# `%` modulo operation.
#
# Difficulty: medium.

def greatest_common_factor(number1, number2):
    greater_number = number1 if number1 >= number2 else number2
    for possible_factor in reversed(range(1, greater_number)):
        if number1 % possible_factor == 0 and number2 % possible_factor == 0:
            return possible_factor

# These are tests to check that your code is working. After writing
# your solution, they should all print True.

print(
  'greatest_common_factor(3, 9) == 3: ' +
  str(greatest_common_factor(3, 9) == 3)
)
print(
  'greatest_common_factor(16, 24) == 8: ' +
  str(greatest_common_factor(16, 24) == 8)
)
print(
  'greatest_common_factor(3, 5) == 1: ' +
  str(greatest_common_factor(3, 5) == 1)
)

File: Ruby/solutions (from jayqui)/18-greatest-common-factor.rb
# Write a method that takes in two numbers. Return the greatest
# integer that evenly divides both numbers. You may wish to use the
# `%` modulo operation.
#
# Difficulty: medium.

def greatest_common_factor(n1, n2)
  candidate_divisor = n1 < n2 ? n1 : n2
  # ^ if this line is confusing, google, 'Ruby ternary operator'.

  while candidate_divisor >= 1
    if n1 % candidate_divisor == 0 && n2 % candidate_divisor == 0
      return candidate_divisor
    end
    candidate_divisor -= 1
  end
end


# These are tests to check that your code is working. After writing
# your solution, they should all print true.

puts(
  'greatest_common_factor(3, 9) == 3: ' +
  (greatest_common_factor(3, 9) == 3).to_s
)
puts(
  'greatest_common_factor(16, 24) == 8: ' +
  (greatest_common_factor(16, 24) == 8).to_s
)
puts(
  'greatest_common_factor(3, 5) == 1: ' +
  (greatest_common_factor(3, 5) == 1).to_s
)
